### ISSUE TITLE
[Feature enhancement] Find and Replace Text

### ISSUE THREAD

**Issue by Aakashkumar22** [Stirling-Tools/Stirling-PDF#1855]
### Feature Description

**Finding and Replacing Text** in a PDF is a feature that allows users to search for specific text within a PDF document and replace it with new text. This can be particularly useful for editing documents, correcting errors, updating information, or making consistent changes throughout a document. Here's a detailed description of how this feature typically works:

**1. Using the Find Function**
Enter Search Term: Type the text you want to find in the search box. This could be a word, phrase, or specific string of characters.
Search Options: Some editors allow you to refine your search by distinguishing between case-sensitive matches, whole words only, or searching within specific sections of the document.

**2. Reviewing Search Results**
Highlighting: The software will highlight all instances of the text you’re searching for throughout the document.
Navigation: Use navigation buttons or arrows to move between occurrences of the found text.

**3. Replacing Text**
Access Replace Option: If available, switch to the "**Replace**" tab or option within the search tool.
Enter Replacement Text: Type the new text you want to replace the old text with.
Replace Options:
Replace: Replace the current highlighted instance with the new text.
Replace All: **Replace all** occurrences of the found text with the new text throughout the entire document.

**4. Confirming Changes**
Review: After replacing text, review the document to ensure that all changes have been made as intended and that the document’s formatting and content remain correct.
Save: Save the edited document by selecting "File" > "Save" or "Save As" to preserve the changes.

### Why is this feature valuable?

The Find and Replace feature is highly valuable in PDF editing for several reasons:

1. **Efficiency and Time-Saving**
Bulk Changes: Instead of manually searching through each page to find specific text, Find and Replace allows you to make bulk changes quickly. This can save significant time, especially in long or complex documents.
Automated Updates: Automating the process of updating information (such as addresses, names, or terms) ensures that changes are made consistently and accurately.
2. **Consistency**
Uniformity: Ensures that changes are applied uniformly across the entire document. This is especially important for maintaining consistent terminology, formatting, or information throughout the document.
Error Reduction: Reduces the risk of human error that can occur when manually updating text. Automated replacements help ensure that all instances of the old text are addressed.
3. **Correction of Errors**
Typographical Errors: Quickly correct typos or spelling mistakes across the document. This is essential for maintaining professionalism and clarity.
Fact Corrections: Fix outdated or incorrect information, such as outdated statistics or references, ensuring that the document reflects current and accurate data.
4. **Document Updates**
Contact Information: Easily update contact details, like phone numbers or email addresses, across multiple documents without needing to edit each instance individually.
Terminology Changes: Modify terminology or jargon that may have become outdated or needs standardization, which is especially useful in technical documents or corporate materials.
5. **Streamlining Content Management**
Content Revision: When revising or editing content, especially in large documents or templates, Find and Replace helps in maintaining coherence and managing content effectively.
Template Updates: For documents based on templates, you can quickly update placeholder text or variable fields to reflect new information or changes.
6. **Document Integrity**
Minimized Errors: Automated replacement helps in minimizing discrepancies that could arise from manual edits, preserving the document’s integrity.
Formatting Preservation: Many Find and Replace tools in PDF editors preserve the document’s formatting and layout, ensuring that changes do not disrupt the visual presentation of the document.
7. **Enhanced Workflow**
Streamlined Editing: Integrating Find and Replace into the editing workflow allows for faster and more streamlined document preparation, editing, and review processes.
Scalability: Useful for handling large volumes of documents that require similar updates, such as legal contracts, company reports, or educational materials.
Real-Life Examples:
**Business Contracts:** If a company changes its address or legal entity name, Find and Replace allows updating these details in all related contracts quickly.
**Academic Papers:** Correcting a term or citation style across a lengthy thesis or research paper ensures consistency and adherence to formatting guidelines.
**Marketing Materials:** Updating promotional materials with new contact information or branding elements can be done efficiently, ensuring all materials reflect current information.
In summary, the Find and Replace feature enhances efficiency, accuracy, and consistency in PDF editing, making it an indispensable tool for managing and updating content across a wide range of documents.

### Suggested Implementation

For the backend Part: 
Business Logic: Implement find and replace logic in a service layer  for string manipulation.
Integration: Expose functionality via a REST controller if needed.
Data Sources: Handle different data sources (files, databases) as required.
Testing: Ensure correctness with appropriate unit tests.

### Additional Information

Overall the find and Replace text can be look like this 

**Before  Text Replacing:**
[
<img width="958" alt="stirling" src="https://github.com/user-attachments/assets/acedb082-96b3-4c3f-8147-bdba275beb58">
](url)

**After  Text Replacing :**
<img width="838" alt="find and replace res" src="https://github.com/user-attachments/assets/15416bb1-0d74-435a-b8b3-2d28f9932018">

### No Duplicate of the Feature

- [X] I have verified that there are no existing features requests similar to my request.

**Comment by Aakashkumar22** [comment]
in the above prototype we can see that **you** is replaced by **I** Everywhere so that kind of functionality we can basically  add in the view Pdf search section  that I can think of according to the  user's prospective.

**Comment by MariaLaranjeira** [comment]
I find this a very good suggestion, but i must ask, what would be your idea for the text replacement itself? Draw a white rectangle over the original text and then draw the new text on top in the same location? @Aakashkumar22

**Comment by Aakashkumar22** [comment]
**"Yes,** that can be stated; however, the text should be written in such a way that it is **entered** **only** **once**,and **all** **instances** where it appears should be **easily replaceable.** depending on the Use case like one time  if he wants it should be updated on the basis of location where he wats to Update it but if he wants to replace  All Instances Then All location Instances should be replaceable. As I mentioned earlier in the supported document, I entered **'You'** only once, which is replaceable by **'I',** and wherever the instance of 'You' is present, it is automatically updated accordingly."

**Comment by Frooodle** [comment]
> I find this a very good suggestion, but i must ask, what would be your idea for the text replacement itself? Draw a white rectangle over the original text and then draw the new text on top in the same location? @Aakashkumar22 


I think just actual removal of the text is what is intended here not drawing ontop 
The issue is when it comes to re writing the text as you may face formatting issues such as replacing a 4 letter word with a 10 letter word and it overlapping other text or going off page etc

**Comment by christaikobo** [comment]
Another case this can be useful is that I sometimes want a removal kind of redaction. 

The current redaction means black box over text, but sometimes there are editable watermarks from some PDFs that I wish to simply remove the texts. 

If I can search these watermarks and replace with nothing, they are removed. Some other tools can do this already.

**Comment by tharun423** [comment]
is it still open or any one working on it as i can see the backend is implemented at /api/v1/general/edit-text but the UI element doesn't include it  @Frooodle  could you please clarify once

**Comment by Frooodle** [comment]
It is being worked on I actually have a new version of text editor in draft at the moment
But no ETA for specifically find replace (and a API version of it)

**Comment by tharun423** [comment]
ok
